File: dynamic-programming/leetcode198.py
#!/usr/bin/env python
# -*- coding:utf-8 -*-
"""
    @Header leetcode198
    @Author fosaken
    @Date   2020/12/17
    @Abstract:
        Address: https://leetcode-cn.com/problems/house-robber/
"""

from typing import List


class Solution:
    # 注意和746的区别
    def rob(self, nums: List[int]) -> int:
        l = len(nums)
        if not l:
            return 0
        if l == 1:
            return nums[0]
        elif l == 2:
            return max(nums[0], nums[1])
        dp = [0] * l
        dp[0] = nums[0]
        dp[1] = nums[1]
        dp[2] = max(nums[0] + nums[2], nums[1])
        for i in range(3, l):
            dp[i] = max(dp[i-3] + nums[i], dp[i-2] + nums[i], dp[i-1])
        return max(dp[-1], dp[-2])


def main():
    solution = Solution()
    nums = [2, 1, 1, 2]
    result = solution.rob(nums)
    print(result)
    pass


if __name__ == '__main__':
    main()

File: dynamic-programming/leetcode198.cpp
#include <iostream>
#include <vector>

using namespace std;

int rob(vector<int>& nums){
    vector<int> counts(nums.size() + 1, 0);
    if (nums.size() == 0){
        return 0;
    }else if(nums.size() == 1){
        return nums[0];
    } else if (nums.size() == 2){
        return max(nums[0], nums[1]);
    }
    counts[1] = nums[0];
    counts[2] = max(nums[0], nums[1]);
    for (int i = 3; i < nums.size()+1; i++){
        counts[i] = max(counts[i - 1], nums[i-1] + counts[i - 2]);
    }
    return counts[nums.size()];
}

int rob1(vector<int>& nums){
    if (nums.size() == 0){
        return 0;
    }else if(nums.size() == 1){
        return nums[0];
    } else if (nums.size() == 2){
        return max(nums[0], nums[1]);
    }
    int a = nums[0];
    int b = max(nums[0], nums[1]);
    int s = 0;
    for (int i = 3; i < nums.size() + 1; i++){
        s = max(b, a + nums[i - 1]);
        a = b;
        b = s;
    }
    return s;
}
int main(){
    vector<int> nums = {2,7,9,3,1};
    cout << rob1(nums);
    system("pause");
    return 0;
}